Src/Plugins/Security/SEC101_050.IdentifiableNpmAuthorTokenValidator.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+
+using Microsoft.CodeAnalysis.Sarif.PatternMatcher.Sdk;
+using Microsoft.RE2.Managed;
+
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher.Plugins.Security
+{
+    public class IdentifiableNpmAuthorTokenValidator : DynamicValidatorBase
+    {
+        internal const string Uri = "https://registry.npmjs.com/-/npm/v1/tokens";
+
+        internal static ValidationState CheckInformation(string content, string secret, ref string message, ref ResultLevelKind resultLevelKind)
+        {
+            TokensRoot tokensRoot = JsonConvert.DeserializeObject<TokensRoot>(content);
+            if (tokensRoot?.Tokens?.Count > 0)
+            {
+                foreach (Object obj in tokensRoot.Tokens)
+                {
+                    if (!secret.Contains(obj.Token))
+                    {
+                        continue;
+                    }
+
+                    if (obj.Readonly)
+                    {
+                        message = "The token has 'read' permissions.";
+                        resultLevelKind = new ResultLevelKind { Level = FailureLevel.Warning };
+                        return ValidationState.Authorized;
+                    }
+
+                    if (obj.Automation)
+                    {
+                        message = "The token has 'automation' permissions.";
+                        return ValidationState.Authorized;
+                    }
+
+                    message = "The token has 'publish' permissions.";
+                    return ValidationState.Authorized;
+                }
+            }
+
+            return ValidationState.Authorized;
+        }
+
+        protected override IEnumerable<ValidationResult> IsValidStaticHelper(IDictionary<string, FlexMatch> groups)
+        {
+            FlexMatch secret = groups["secret"];
```

Comment:
>secret [](http://example.com/codeflow?start=22&length=6)

to reduce false-positives, we should validate the checksum:
[https://github.blog/changelog/2021-09-23-npm-has-a-new-access-token-format/](https://github.blog/changelog/2021-09-23-npm-has-a-new-access-token-format/)

Reply:
Check sec101/102 or sec101/006 (internal version)